src/fu-dbus-daemon.c

```diff
@@ -1635,19 +1722,31 @@ fu_dbus_daemon_method_modify_device(FuDbusDaemon *self,
 				    FuEngineRequest *request,
 				    GDBusMethodInvocation *invocation)
 {
-	FuEngine *engine = fu_daemon_get_engine(FU_DAEMON(self));
 	const gchar *device_id;
 	const gchar *key = NULL;
 	const gchar *value = NULL;
-	g_autoptr(GError) error = NULL;
+	g_autoptr(FuMainAuthHelper) helper = NULL;
 
-	/* check the id exists */
 	g_variant_get(parameters, "(&s&s&s)", &device_id, &key, &value);
-	if (!fu_engine_modify_device(engine, device_id, key, value, &error)) {
-		fu_dbus_daemon_method_invocation_return_gerror(invocation, error);
-		return;
-	}
-	g_dbus_method_invocation_return_value(invocation, NULL);
+	helper = g_new0(FuMainAuthHelper, 1);
+	helper->self = self;
+	helper->request = g_object_ref(request);
+	helper->invocation = g_object_ref(invocation);
+	helper->device_id = g_strdup(device_id);
+	helper->key = g_strdup(key);
+	helper->value = g_strdup(value);
+	if (fu_dbus_daemon_method_modify_device_flag_needs_auth(key, value)) {
+		fu_dbus_daemon_set_status(self, FWUPD_STATUS_WAITING_FOR_AUTH);
+		fu_polkit_authority_check(
+		    self->authority,
+		    fu_engine_request_get_sender(request),
+		    "org.freedesktop.fwupd.emulation-tag",
+		    fu_dbus_daemon_engine_request_get_authority_check_flags(request),
+		    NULL,
+		    fu_dbus_daemon_authorize_modify_device_cb,
+		    g_steal_pointer(&helper));
+	} else
+		fu_dbus_daemon_authorize_modify_device_cb(NULL, NULL, g_steal_pointer(&helper));
```

Comment:
I don't really like this stylistically. Can we do a fu_dbus_daemon_authorize_modify_device_internal or something that we call here and from fu_dbus_daemon_authorize_modify_device_cb please.

Reply:
fixed